service/worker/scanner/shardscanner/fixer_workflow.go

```diff
@@ -0,0 +1,226 @@
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package shardscanner
+
+import (
+	"errors"
+
+	"go.uber.org/cadence/workflow"
+
+	"github.com/uber/cadence/common/blobstore"
+	"github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/reconciliation/invariant"
+	"github.com/uber/cadence/common/reconciliation/store"
+)
+
+const (
+	fixShardReportChan = "fixShardReportChan"
+)
+
+var (
+	errQueryNotReady = errors.New("query is not yet ready to be handled, please try again shortly")
+)
+
+// FixerManagerCB is a function which returns invariant manager for fixer.
+type FixerManagerCB func(
+	persistence.Retryer,
+	FixShardActivityParams,
+	ScannerConfig,
+) invariant.Manager
+
+// FixerIteratorCB is a function which returns ScanOutputIterator for fixer.
+type FixerIteratorCB func(
+	blobstore.Client,
+	store.Keys,
+	FixShardActivityParams,
+	ScannerConfig,
+) store.ScanOutputIterator
+
+// FixerHooks holds callback functions for shard scanner workflow implementation.
+type FixerHooks struct {
+	InvariantManager FixerManagerCB
+	Iterator         FixerIteratorCB
+}
+
+// NewFixerHooks returns initialized callbacks for shard scanner workflow implementation.
+func NewFixerHooks(manager FixerManagerCB, iterator FixerIteratorCB) (*FixerHooks, error) {
+	if manager == nil || iterator == nil {
+		return nil, errors.New("manager or iterator not provided")
+	}
+	return &FixerHooks{
+		InvariantManager: manager,
+		Iterator:         iterator,
+	}, nil
+}
+
+// FixerWorkflow is the workflow that fixes all entities from a scan output.
+type FixerWorkflow struct {
+	Aggregator *ShardFixResultAggregator
+	Params     FixerWorkflowParams
+	Keys       *FixerCorruptedKeysActivityResult
+}
+
+// NewFixerWorkflow returns a new instance of fixer workflow
+func NewFixerWorkflow(
+	ctx workflow.Context,
+	name string,
+	params FixerWorkflowParams,
+) (*FixerWorkflow, error) {
+
+	if len(name) < 1 {
+		return nil, errors.New("workflow name is not provided")
+	}
+
+	params.ContextKey = ScannerContextKey(name)
+
+	wf := FixerWorkflow{
+		Params: params,
+	}
+
+	corruptKeys, err := GetCorruptedKeys(ctx, wf.Params)
+	if err != nil {
+		return nil, err
+	}
+	wf.Keys = corruptKeys
+	wf.Aggregator = NewShardFixResultAggregator(corruptKeys.CorruptedKeys, *corruptKeys.MinShard, *corruptKeys.MaxShard)
+
+	for name, fn := range setHandlers(wf.Aggregator) {
+		if err := workflow.SetQueryHandler(ctx, name, fn); err != nil {
```

Comment:
So are different scan types able to set different types of queries? For example what if I want a query on timers that says how many timers are going to fire between time X and Y?

Reply:
Yes, we are on the same page here. (the same applies to metrics part) I will make those parts extendable in a separate PR.